File: lda/importance_sampling.m
function log_evidence = ldae_is_variants(words, topics, topic_prior, num_samples, variant, variant_iters)
%LDAE_IS_VARIANTS approximate evidence of LDA model, importance sampling with choice of q-distributions
%
% log_evidence = ldae_is_variants(words, topics, topic_prior[, num_samples=1000[, variant=3[, variant_iters=1]]]);
%
% Inputs:
%             words 1xNd
%            topics TxV each row is a distribution over a vocabulary of size V 
%       topic_prior 1xT parameters of Dirichlet from which document topic vector is drawn
%       num_samples 1x1 default 1000
%           variant 1x1 1: prior, 2: q(z) = \prod p(z_n|w_n), 3: q(z) = \prod(z_n|w_n,hacky_pseudo_counts)
%                       Default is 3, as that seems to work best.
%     variant_iters 1x1 If variant needs iterative updates, use this number
%
% Outputs:
%     log_evidence  1x1 

% Iain Murray, January 2009

[T, V] = size(topics);
Nd = length(words);

if ~exist('num_samples', 'var')
    num_samples = 1000;
end
if ~exist('variant', 'var')
    variant = 3;
end
if ~exist('variant_iters', 'var')
    variant_iters = 1;
end

// The dimension of the new matrix is T X 1
// equivalent to alpha * m
topic_prior = topic_prior(:);

topic_alpha = sum(topic_prior); 
%topic_mean = topic_prior / topic_alpha;

if variant == 1
    % Importance sample from prior
    qstar = repmat(topic_prior, 1, Nd); % T x Nd 
    qq = bsxfun(@rdivide, qstar, sum(qstar, 1));  // normalize
else
    //Take w_n into account when picking z_n
    qstar = bsxfun(@times, topic_prior, topics(:, words)); % T x Nd // phi, alpha*m
    qq = bsxfun(@rdivide, qstar, sum(qstar, 1)); // normalize

    if variant == 3
        for i = 1:variant_iters
            % Now create pseudo-counts from qq and recompute qq using them
            pseudo_counts = bsxfun(@minus, topic_prior + sum(qq, 2), qq); // finding the pseudo counts
            qstar = bsxfun(@times, pseudo_counts, topics(:, words)); % T x Nd
            qq = bsxfun(@rdivide, qstar, sum(qstar, 1));
        end
    end
end

% Draw samples from the q-distribution
samples = zeros(Nd, num_samples); // Nd X 1000
for n = 1:Nd // for every word
    samples(n, :) = discreternd(num_samples, qq(:, n)); % Nd x num_samples // sample 1000 times from some discrete distribution for one word // samples the topic for corresponding to a word
end 

// Evaluate P(z, v) at samples and compare to q-distribution
// no of samples belonging to each topic.
Nk = histc(samples, 1:T, 1); // T x num_samples
// log(gamma(T X No_samples + T X 1)) + log(gamma(topic_alpha)) - log(gamma(topic_prior)) - log(gamma(Nd + topic_alpha))
// some normalization constant?
log_pz = sum(gammaln(bsxfun(@plus, Nk, topic_prior)), 1) ...
        + gammaln(topic_alpha) - sum(gammaln(topic_prior)) ...
        - gammaln(Nd + topic_alpha); // 1 x num_samples
log_w_given_z = zeros(1, num_samples); // P(w/z)
for n = 1:Nd // for each word, select all the samples * each word
    // P(w/z) = table lookup
    // product of P(w/z) = sum(log(P(w/z)))
    log_w_given_z = log_w_given_z + log(topics(samples(n,:), words(n)));
end
// log(P(w/z)) + log(P(z))
log_joint = log_pz + log_w_given_z;

%

// 1*num_samples
log_qq = zeros(1, num_samples);
for n = 1:Nd // sample only the topic for the dividor
    log_qq = log_qq + log(qq(samples(n,:), n));
end
log_weights = log_joint - log_qq;
// equivalent to normalizing over all the samples
// log(exp(s1) + exp(s2) + exp(s1000))
log_evidence = logsumexp(log_weights(:)) - log(length(log_weights));

File: lda/importance_sampling.py
import numpy as np
import math
from scipy.special import loggamma, logsumexp

def importance_sampling(document, topic_word, topic_prior, num_samples, topic_list):
	'''
		document - document of words in the form a list 
		topic_word - Topic word distribution learned from a model. (T X V)
		topic_prior - Dirichlet distribution from which document topic vector is drawn. (1 X T)
		num_samples - No of samples
		topic_list - List of topics
	'''

	# number of words in the document
	Nd = len(document)

	# 1. change the topic priors to Tx1
	topic_prior = topic_prior.transpose()
	T = topic_prior.shape[0] # T - number of topics

	# 2. find the topic alpha. will be used to find the P(Z)
	# T * topic_prior
	topic_alpha = topic_prior.sum()

	# 3. Importance sampling from prior
	# Proposal distribution T X Nd
	qq = np.repeat(topic_prior, Nd, axis=1)
	qq = qq / qq.sum(axis=0) # probability of word summed over all topics

	# 4. draw samples from the discrete distribution
	# sample topics from a discrete distribution for each word.
	# Nd X num_samples
	samples = np.zeros((Nd, num_samples), dtype=np.uint8)

	for n in range(Nd):
		samples[n] = np.random.choice(topic_list, num_samples, list(qq[:,n])) # in this case, the probability will be the same.


	# 5. Evaluate P(z, w) from the samples and compare it with q-distribution
	# number of times a word from document d has been assigned to topic t. (for each sample how many words belong to a particular topic. )
	Nk = np.zeros((T, num_samples), dtype=np.uint8)		# T X num_samples
	for n in range(num_samples):
		Nk[:,n] = np.bincount(samples[:,n], minlength=T)
	
	# P(z)
	log_pz = loggamma(Nk+topic_prior).sum(axis=0) \
			+ loggamma(topic_alpha) - loggamma(topic_prior).sum() \
			- loggamma(Nd+topic_alpha)
	# log_pz = math.log(math.gamma(Nk+topic_prior)).sum(axis=0) \
	# 		+ math.log(math.gamma(topic_alpha)) - math.log(math.gamma(topic_prior)).sum() \
	# 		- math.log(math.gamma(Nd+topic_alpha))

	# P(w/z)
	log_w_given_z = np.zeros((1, num_samples), dtype=np.float64)
	for n in range(Nd):
		log_w_given_z += np.log(topic_word[samples[n], document[n]]).reshape(1, -1)

	# find the joint distribution P(z, w) = P(z) * P(w/z)
	log_joint = log_w_given_z + log_pz # 1 X num_samples

	# 6. find the distribution P(z) (1 X num_samples)
	log_qq = np.zeros((1, num_samples), dtype=np.float64)
	for n in range(Nd):
		log_qq += np.log(qq[samples[n], n]).reshape(1, -1)

	# 7. find the probability of the document.
	# P(w) = P(w, z) / P(z)
	log_weight = log_joint - log_qq

	assert(log_weight.shape[1], num_samples)

	# normalize for the number of samples
	log_evidence = logsumexp(log_weight) - math.log(num_samples)

	return log_evidence

